File: neet_code_150/src/array_hash/valid_sudoku.rs
use std::collections::{HashMap, HashSet};

struct Solution {}
impl Solution {
    // 解けなかった😭
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        for i in 0..9 {
            let mut number_set = HashSet::new();
            for j in 0..9 {
                if board[i][j] != '.' {
                    number_set.insert(board[i][j]);
                }
            }
        }
        false
    }

    // C++のpassする模範解答をRustに変換した
    pub fn is_valid_sudoku_2(board: Vec<Vec<char>>) -> bool {
        const CNT: usize = 9;
        let mut row: [[bool; CNT]; CNT] = [[false; CNT]; CNT];
        let mut col: [[bool; CNT]; CNT] = [[false; CNT]; CNT];
        let mut sub: [[bool; CNT]; CNT] = [[false; CNT]; CNT];

        for r in 0..CNT {
            for c in 0..CNT {
                // 数字でないならpass
                if board[r][c] == '.' {
                    continue;
                }

                // char('1' ~ '9') -> index
                let index = board[r][c] as usize - '0' as usize - 1;
                let area = (r / 3) * 3 + (c / 3);

                // if nubmer already exists
                if row[r][index] || col[c][index] || sub[area][index] {
                    return false;
                }

                row[r][index] = true;
                col[c][index] = true;
                sub[area][index] = true;
            }
        }
        true
    }
}

struct SolutionAns {}
impl SolutionAns {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut row = HashSet::new();
        let mut col = HashSet::new();
        let mut bx = HashSet::new();

        for i in 0..9 {
            for j in 0..9 {
                let r = board[i][j];
                let c = board[j][i];
                let b = board[i / 3 * 3 + j / 3][i % 3 * 3 + j % 3];

                if r != '.' {
                    if !row.insert(r) {
                        return false;
                    }
                }

                if c != '.' {
                    if !col.insert(c) {
                        return false;
                    }
                }

                if b != '.' {
                    if !bx.insert(b) {
                        return false;
                    }
                }
            }

            row.clear();
            col.clear();
            bx.clear();
        }
        true
    }
}

// AC
// Pythonの模範解答より
struct SolutionAnsPython {}
impl SolutionAnsPython {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        // key = r, key = c
        let mut row: HashMap<i32, HashSet<char>> = HashMap::new();
        let mut col: HashMap<i32, HashSet<char>> = HashMap::new();
        // key = (r/3, c/3)
        let mut subcell: HashMap<(i32, i32), HashSet<char>> = HashMap::new();

        for r in 0..9_i32 {
            for c in 0..9_i32 {
                if board[r as usize][c as usize] == '.' {
                    continue;
                }
                let b_rc = board[r as usize][c as usize];

                for opt in [row.get(&r), col.get(&c), subcell.get(&(r / 3, c / 3))] {
                    if let Some(opt_raw) = opt {
                        if opt_raw.contains(&b_rc) {
                            return false;
                        }
                    }
                }

                row.entry(r).or_insert(HashSet::from([b_rc])).insert(b_rc);
                col.entry(c).or_insert(HashSet::from([b_rc])).insert(b_rc);
                subcell
                    .entry((r / 3, c / 3))
                    .or_insert(HashSet::from([b_rc]))
                    .insert(b_rc);
            }
        }

        true
    }
}

// 後から解いたときの別解
struct SolutionLatter;
impl SolutionLatter {
    // AC
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut set_h = vec![HashSet::new(); 9];
        let mut set_v = vec![HashSet::new(); 9];
        let mut set_c = vec![vec![HashSet::new(); 3]; 3];

        for a in 0..9 {
            for b in 0..9 {
                let c = board[a][b];
                if c == '.' {
                    continue;
                }

                if set_h[a].contains(&c)
                    || set_v[b].contains(&c)
                    || set_c[a / 3][b / 3].contains(&c)
                {
                    return false;
                } else {
                    set_h[a].insert(c);
                    set_v[b].insert(c);
                    set_c[a / 3][b / 3].insert(c);
                }
            }
        }

        true
    }
}

fn main() {
    let case_1 = vec![
        vec!['5', '3', '.', '.', '7', '.', '.', '.', '.'],
        vec!['6', '.', '.', '1', '9', '5', '.', '.', '.'],
        vec!['.', '9', '8', '.', '.', '.', '.', '6', '.'],
        vec!['8', '.', '.', '.', '6', '.', '.', '.', '3'],
        vec!['4', '.', '.', '8', '.', '3', '.', '.', '1'],
        vec!['7', '.', '.', '.', '2', '.', '.', '.', '6'],
        vec!['.', '6', '.', '.', '.', '.', '2', '8', '.'],
        vec!['.', '.', '.', '4', '1', '9', '.', '.', '5'],
        vec!['.', '.', '.', '.', '8', '.', '.', '7', '9'],
    ];

    let case_2 = vec![
        vec!['8', '3', '.', '.', '7', '.', '.', '.', '.'],
        vec!['6', '.', '.', '1', '9', '5', '.', '.', '.'],
        vec!['.', '9', '8', '.', '.', '.', '.', '6', '.'],
        vec!['8', '.', '.', '.', '6', '.', '.', '.', '3'],
        vec!['4', '.', '.', '8', '.', '3', '.', '.', '1'],
        vec!['7', '.', '.', '.', '2', '.', '.', '.', '6'],
        vec!['.', '6', '.', '.', '.', '.', '2', '8', '.'],
        vec!['.', '.', '.', '4', '1', '9', '.', '.', '5'],
        vec!['.', '.', '.', '.', '8', '.', '.', '7', '9'],
    ];

    println!("case_1: {}", Solution::is_valid_sudoku_2(case_1.clone()));
    println!("case_2: {}", Solution::is_valid_sudoku_2(case_2.clone()));

    println!("case_1: {}", SolutionAns::is_valid_sudoku(case_1.clone()));
    println!("case_2: {}", SolutionAns::is_valid_sudoku(case_2.clone()));

    println!(
        "case_1: {}",
        SolutionAnsPython::is_valid_sudoku(case_1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAnsPython::is_valid_sudoku(case_2.clone())
    );

    println!(
        "case_1: {}",
        SolutionLatter::is_valid_sudoku(case_1.clone())
    );
    println!(
        "case_2: {}",
        SolutionLatter::is_valid_sudoku(case_2.clone())
    );
}

File: neet_code_150_cpp/array_hash/valid_sudoku.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <map>

// 模範解答
class SolutionAns {
    public:
        bool is_valid_sudoku(const std::vector<std::vector<char>> & board) {
            const int cnt = 9;
            bool row[cnt][cnt] = {false};
            bool col[cnt][cnt] = {false};
            bool sub[cnt][cnt] = {false};

            for(int r = 0; r < cnt; ++r){
                for(int c = 0; c < cnt; ++c){
                    if(board[r][c] == '.')
                        continue; // if not number pass

                    int idx = board[r][c] - '0' - 1; //char to num idx
                    int area = (r/3) * 3 + (c/3); // これが??

                    //if number already exists
                    if(row[r][idx] || col[c][idx] || sub[area][idx]){
                        return false;
                    }

                    row[r][idx] = true;
                    col[c][idx] = true;
                    sub[area][idx] = true;
                }
            }
            return true;
        }
};

// AC
// Rustの模範解答より
class SolutionAnsRust {
    public:
        bool isValidSudoku(const std::vector<std::vector<char>>& board) {
            std::unordered_set<char> row, col, bx;

            for(int i = 0; i < 9; ++i) {
                for(int j = 0; j < 9; ++j) {
                    auto r = board[i][j];
                    auto c = board[j][i];
                    auto b = board[i / 3 * 3 + j / 3][i % 3 * 3 + j % 3];

                    if(r != '.') {
                        if(row.find(r) != row.end()) {
                            return false;
                        }
                        row.insert(r);
                    }

                    if(c != '.') {
                        if(col.find(c) != col.end()) {
                            return false;
                        }
                        col.insert(c);
                    }

                    if(b != '.') {
                        if(bx.find(b) != bx.end()) {
                            return false;
                        }
                        bx.insert(b);
                    }
                }

                row.clear();
                col.clear();
                bx.clear();
            }

            return true;
        }
};

// AC
// Pythonの模範解答より
class SolutionAnsPython {
    public:
        bool isValidSudoku(const std::vector<std::vector<char>>& board) {
            // key = r, key = c
            std::unordered_map<int, std::unordered_set<char>> row, col;
            // key = {r / 3, c /3}
            std::map<std::pair<int,int>, std::unordered_set<char>> subcell;

            for(int r = 0; r < 9; ++r) {
                for(int c = 0; c < 9; ++c) {
                    if(board[r][c] == '.') {
                        continue;
                    }

                    auto b_rc = board[r][c];
                    if(row[r].find(b_rc) != row[r].end()
                            || col[c].find(b_rc) != col[c].end()
                            || subcell[std::pair(r/3, c/3)].find(b_rc) 
                            != subcell[std::pair(r/3, c/3)].end() 
                            ) {
                        return false;
                    }

                    row[r].insert(b_rc);
                    col[c].insert(b_rc);
                    subcell[std::pair(r/3, c/3)].insert(b_rc);
                }
            }

            return true;
        }
};

int main(void) {
     auto case_1 = std::vector{
        std::vector{'5', '3', '.', '.', '7', '.', '.', '.', '.'},
        std::vector{'6', '.', '.', '1', '9', '5', '.', '.', '.'},
        std::vector{'.', '9', '8', '.', '.', '.', '.', '6', '.'},
        std::vector{'8', '.', '.', '.', '6', '.', '.', '.', '3'},
        std::vector{'4', '.', '.', '8', '.', '3', '.', '.', '1'},
        std::vector{'7', '.', '.', '.', '2', '.', '.', '.', '6'},
        std::vector{'.', '6', '.', '.', '.', '.', '2', '8', '.'},
        std::vector{'.', '.', '.', '4', '1', '9', '.', '.', '5'},
        std::vector{'.', '.', '.', '.', '8', '.', '.', '7', '9'},
     };
    auto case_2 = std::vector{
        std::vector{'8', '3', '.', '.', '7', '.', '.', '.', '.'},
        std::vector{'6', '.', '.', '1', '9', '5', '.', '.', '.'},
        std::vector{'.', '9', '8', '.', '.', '.', '.', '6', '.'},
        std::vector{'8', '.', '.', '.', '6', '.', '.', '.', '3'},
        std::vector{'4', '.', '.', '8', '.', '3', '.', '.', '1'},
        std::vector{'7', '.', '.', '.', '2', '.', '.', '.', '6'},
        std::vector{'.', '6', '.', '.', '.', '.', '2', '8', '.'},
        std::vector{'.', '.', '.', '4', '1', '9', '.', '.', '5'},
        std::vector{'.', '.', '.', '.', '8', '.', '.', '7', '9'},
    };

    SolutionAns s_1;
    std::cout <<"case_1" << std::endl;
    if(s_1.is_valid_sudoku(case_1)) {
        std::cout << "valid sudoku" << std::endl;
    }
    std::cout <<"case_2" << std::endl;
    if(s_1.is_valid_sudoku(case_2)) {
        std::cout << "valid sudoku" << std::endl;
    }

    SolutionAnsRust s_ans_rs;
    std::cout << s_ans_rs.isValidSudoku(case_1) << std::endl;
    std::cout << s_ans_rs.isValidSudoku(case_2) << std::endl;

    SolutionAnsPython s_ans_py;
    std::cout << s_ans_py.isValidSudoku(case_1) << std::endl;
    std::cout << s_ans_py.isValidSudoku(case_2) << std::endl;
}
